common/flatpak-dir.c

```diff
@@ -8909,6 +8909,188 @@ flatpak_dir_deploy_update (FlatpakDir        *self,
   return TRUE;
 }
 
+static void
+rewrite_one_dynamic_launcher (const char *portal_desktop_dir,
+                              const char *portal_icon_dir,
+                              const char *desktop_name,
+                              const char *old_app_id,
+                              const char *new_app_id)
+{
+  g_autoptr(GKeyFile) old_key_file = NULL;
+  g_autoptr(GKeyFile) new_key_file = NULL;
+  g_autoptr(GFile) desktop_file = NULL;
+  g_autoptr(GFile) link_file = NULL;
+  g_autoptr(GFile) new_link_file = NULL;
+  g_autoptr(GString) data_string = NULL;
+  g_autoptr(GError) local_error = NULL;
+  g_autofree char *old_data = NULL;
+  g_autofree char *desktop_path = NULL;
+  g_autofree char *icon_path = NULL;
+  g_autofree char *relative_path = NULL;
+  g_autofree char *new_desktop = NULL;
+  const gchar *desktop_suffix;
+
+  g_assert (g_str_has_suffix (desktop_name, ".desktop"));
+
+  desktop_path = g_build_filename (portal_desktop_dir,
+                                   desktop_name, NULL);
+  old_key_file = g_key_file_new ();
+  if (!g_key_file_load_from_file (old_key_file, desktop_path,
+                                  G_KEY_FILE_KEEP_COMMENTS | G_KEY_FILE_KEEP_TRANSLATIONS,
+                                  &local_error))
+    {
+      g_warning ("Error encountered loading key file %s: %s", desktop_path, local_error->message);
+      return;
+    }
+  if (!g_key_file_has_key (old_key_file, G_KEY_FILE_DESKTOP_GROUP, "X-Flatpak", NULL))
+    {
+      g_debug ("Ignoring non-Flatpak dynamic launcher: %s", desktop_path);
+      return;
+    }
+
+  /* Fix paths in desktop file with a find-and-replace. The portal handled
+   * quoting the app ID in the Exec line for us.
+   */
+  old_data = g_key_file_to_data (old_key_file, NULL, NULL);
+  data_string = g_string_new ((const char *)old_data);
+  g_string_replace (data_string, old_app_id, new_app_id, 0);
+  new_key_file = g_key_file_new ();
+  if (!g_key_file_load_from_data (new_key_file, data_string->str, -1,
+                                  G_KEY_FILE_KEEP_COMMENTS | G_KEY_FILE_KEEP_TRANSLATIONS,
+                                  &local_error))
+    {
+      g_warning ("Cannot load desktop file %s after rewrite: %s", desktop_path, local_error->message);
+      g_warning ("Key file contents:\n%s\n", (const char *)data_string->str);
+      return;
+    }
+  if (!g_key_file_save_to_file (new_key_file, desktop_path, &local_error))
```

Comment:
Would it make more sense to write the `new_key_file` out to `new_desktop` rather than the old `desktop_path`? This would avoid the need for a rename, instead requiring a delete of the old file, which could happen after switching the symlink. That would leave the file system in perhaps a better state if this function fails and returns early at some point.

Currently if it returns early, it’s possible the key file will exist at the old path, but with rewritten contents which don’t match its name; that seems bad (but I haven’t dug fully into the semantics of all these files and whether it actually really matters).

Reply:
Yes, that would be better, good point.